parsers/US_ERCOT.py

```diff
@@ -0,0 +1,208 @@
+"""Parser for the ERCOT grid area of the United States."""
+
+
+import gzip
+import json
+from datetime import datetime
+from logging import Logger, getLogger
+from typing import Optional
+
+import arrow
+import pytz
+from requests import Response, Session
+
+import parsers.EIA as EIA
+from parsers.lib.validation import validate_exchange
+
+TX_TZ = pytz.timezone("US/Central")
+US_PROXY = "https://us-ca-proxy-jfnx5klx2a-uw.a.run.app"
+HOST_PARAMETER = "host=https://www.ercot.com"
+
+RT_GENERATION_URL = (
+    f"{US_PROXY}/api/1/services/read/dashboards/fuel-mix.json?{HOST_PARAMETER}"
+)
+RT_CONSUMPTION_URL = f"{US_PROXY}/api/1/services/read/dashboards/loadForecastVsActual.json?{HOST_PARAMETER}"
+RT_EXCHANGE_URL = (
+    f"{US_PROXY}/api/1/services/read/dashboards/rtsyscond.json?{HOST_PARAMETER}"
+)
+RT_PRICES_URL = (
+    f"{US_PROXY}/api/1/services/read/dashboards/systemWidePrices.json?{HOST_PARAMETER}"
+)
+
+GENERATION_MAPPING = {
+    "Coal and Lignite": "coal",
+    "Hydro": "hydro",
+    "Natural Gas": "gas",
+    "Nuclear": "nuclear",
+    "Other": "unknown",
+    "Power Storage": "battery",
+    "Solar": "solar",
+    "Wind": "wind",
+}
+EXCHANGE_MAPPING = {"US-CENT-SWPP": ["dcE", "dcN"], "MX-NE": ["dcL"], "MX-NO": ["dcR"]}
+
+
+def get_data(url: str, session: Session):
+    """requests ERCOT url and return json"""
+    r: Response = session.get(url)
+    response_text = gzip.decompress(r.content).decode("utf-8")
+    data_json = json.loads(response_text)
+    return data_json
+
+
+def fetch_live_consumption(
+    zone_key: str,
+    session: Session,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    data_json = get_data(url=RT_CONSUMPTION_URL, session=session)
+    all_data_points = []
+    for key in ["previousDay", "currentDay"]:
+        data_dict = data_json[key]
+        dt = arrow.get(data_dict["dayDate"]).datetime.replace(tzinfo=TX_TZ)
+        for item in data_dict["data"]:
+            if "systemLoad" in item:
+                data_point = {
+                    "datetime": dt.replace(hour=item["hourEnding"] - 1),
+                    "consumption": item["systemLoad"],
+                    "zoneKey": zone_key,
+                    "source": "ercot.com",
+                }
+                all_data_points.append(data_point)
+    return all_data_points
+
+
+def fetch_live_production(
+    zone_key: str,
+    session: Session,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    data_json = get_data(url=RT_GENERATION_URL, session=session)["data"]
+    data_dict = {}
+    for key in data_json:
+        data_dict = {**data_dict, **data_json[key]}
+    all_data_points = []
+    for item in data_dict:
+        production = {}
+        storage = {}
+        dt = arrow.get(item).datetime.replace(tzinfo=TX_TZ)
+        for mode in data_dict[item]:
+            value = data_dict[item][mode]["gen"]
+            if mode == "Power Storage":
+                storage[GENERATION_MAPPING[mode]] = value
```

Comment:
Keep in mind that we treat battery values pretty werid and negative values are discharge and positive ones charge. I fell into this trap in https://github.com/electricitymaps/electricitymaps-contrib/pull/4965

So I think this should be:
```suggestion
                storage[GENERATION_MAPPING[mode]] = value * -1
```

Documentation from the example parser:
```python
    for item in obj["storage"]:
        # Positive storage means energy is stored
        # Negative storage means energy is generated from the storage system
        data["storage"][item["key"]] = item["value"]  # Should be a floating point value
```

Reply:
Actually i'm wondering if we should aggregate storage here as unknown to be consistent with the EIA data? from analysing the 2 data sources i think that this share goes in unknown in the EIA.

Reply:
I would say we should put it in storage because of 3 main reasons:

1. We should strive to have as little unknown data as possible and (battery) storage will only grow.
2. Having it separate would allow us to change how we calculate storage emissions in the future without re-fetching a bunch of data. Right now they are inherently flawed as they use a yearly average. (I have some ideas around this but that's another issue).
3. It would allow us to show that battery storage is actually used to balance the grid in the app.

Reply:
in this case should we also make sure that when we backfill with EIA data, this battery storage is not double counted in unknown? Not sure how to proceed here though

Reply:
Oh right, so for compatibility sake it might be better to just put it in the unknown data anyway as double counting it would invalidate my above points anyway. 😅

The only other solution would be if we where able to "deduct" the battery values from the unknown parts of the EIA data but I thinks that's something we would have to do in the backend and would most likely require bigger changes.

Reply:
> Oh right, so for compatibility sake it might be better to just put it in the unknown data anyway as double counting it would invalidate my above points anyway. 😅
> 
> The only other solution would be if we where able to "deduct" the battery values from the unknown parts of the EIA data but I thinks that's something we would have to do in the backend and would most likely require bigger changes. (This would have massive benefitss for other zones as well, such as SE-SE4 where we could use the entso-e per plant data to get oil production but it's published on a 5 day delay)

Reply:
yes maybe we could apply a very simple estimation after with a simple condition like if source==eia.org then split unknown as unknown + battery storage. I'll discuss this with Robin and we'll see next week :)

Reply:
As discussed with Mathilde, since this is a very little generation, for now we agreed that putting it in unknown is easier. With our current code base, if we refetch historical data, we should introduce a static attribution for this zone that a specific share of the unknown goes towards battery storage. Which might be wrong from time to time. Therefore, it's probably more correct to always put it in unknown for now until we can build something better.

Reply:
Sounds good, I'll change my review once the changes have been made.